File: 백준/Silver/9012. 괄호/괄호.cc
#include<bits/stdc++.h>
using namespace std;
int n;
string s;
string tp;
int main()
{
	cin>>n;
	for(int i=0;i<n;i++)
	{
		cin>>s;
		stack<string> stk;
		stk.push(string(1,s[0]));
		for(int i=1;i<s.size();i++)
		{
			if(!stk.empty() && stk.top() + string(1,s[i]) == "()")
			{
				stk.pop();
			}
			else stk.push(string(1,s[i]));
		}
		if(stk.size()==0) cout<<"YES"<<"\n";
		else cout<<"NO"<<"\n";
	}
}

File: 백준/Silver/9012. 괄호/괄호.py
n=int(input().strip())
for _ in range(n):
    line = input().strip()
    stack=[]
    for ch in line:
        if stack:
            if stack[-1] == '(' and ch==')':
                stack.pop()
            else:
                stack.append(ch)
        else:
            stack.append(ch)
    if not stack:
        print('YES')
    else:
        print('NO')